[CI] SearchWhileRelocatingIT testSearchAndRelocateConcurrentlyRandomReplicas failing

Build scans:
https://gradle-enterprise.elastic.co/s/2inrqyqdkzmac
https://gradle-enterprise.elastic.co/s/cwnhluamqffww
https://gradle-enterprise.elastic.co/s/4jh5t7vzqusqm

**Build scan:**
https://gradle-enterprise.elastic.co/s/4jh5t7vzqusqm/tests/:server:internalClusterTest/org.elasticsearch.search.basic.SearchWhileRelocatingIT/testSearchAndRelocateConcurrentlyRandomReplicas

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.basic.SearchWhileRelocatingIT.testSearchAndRelocateConcurrentlyRandomReplicas" -Dtests.seed=63AD7793D7B01433 -Dtests.locale=en-US -Dtests.timezone=SystemV/MST7MDT -Druntime.java=8`

**Applicable branches:**
7.13, master

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.search.basic.SearchWhileRelocatingIT&tests.test=testSearchAndRelocateConcurrentlyRandomReplicas

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=6713, name=Thread-65, state=RUNNABLE, group=TGRP-SearchWhileRelocatingIT]


  Caused by: java.lang.AssertionError: Count is 14 hits but 15 was expected.  Total shards: 20 Successful shards: 18 & 2 shard failures:
   shard [[CkP3Ry9VQwu1C2cLCANqpQ][test][18]], reason [NoShardAvailableActionException[[node_s1][127.0.0.1:56781][indices:data/read/search[phase/query]]]], cause [[test/hy3EWpL-SviYScv0yfpqSg][[test][18]] NoShardAvailableActionException[[node_s1][127.0.0.1:56781][indices:data/read/search[phase/query]]]
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:500)
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:449)
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction.access$000(AbstractSearchAsyncAction.java:62)
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:316)
  	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:66)
  	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)
  	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:400)
  	at org.elasticsearch.transport.TransportService$5.handleException(TransportService.java:738)
  	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1283)
  	at org.elasticsearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:317)
  	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:215)
  	at org.elasticsearch.transport.InboundHandler.handleException(InboundHandler.java:315)
  	at org.elasticsearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:307)
  	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:126)
  	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:84)
  	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:693)
  	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:129)
  	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:104)
  	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:69)
  	at org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:311)
  	at org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:217)
  	at org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:29)
  	at org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
  	at org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:140)
  	at org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:409)
  	at org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:235)
  	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:163)
  	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:120)
  	at java.lang.Thread.run(Thread.java:748)
  ]
   shard [[CkP3Ry9VQwu1C2cLCANqpQ][test][19]], reason [NoShardAvailableActionException[[node_s1][127.0.0.1:56781][indices:data/read/search[phase/query]]]], cause [[test/hy3EWpL-SviYScv0yfpqSg][[test][19]] NoShardAvailableActionException[[node_s1][127.0.0.1:56781][indices:data/read/search[phase/query]]]
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:500)
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:449)
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction.access$000(AbstractSearchAsyncAction.java:62)
  	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:316)
  	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:66)
  	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)
  	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:400)
  	at org.elasticsearch.transport.TransportService$5.handleException(TransportService.java:738)
  	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1283)
  	at org.elasticsearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:317)
  	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:215)
  	at org.elasticsearch.transport.InboundHandler.handleException(InboundHandler.java:315)
  	at org.elasticsearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:307)
  	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:126)
  	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:84)
  	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:693)
  	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:129)
  	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:104)
  	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:69)
  	at org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:311)
  	at org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:217)
  	at org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:29)
  	at org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
  	at org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:140)
  	at org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:409)
  	at org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:235)
  	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:163)
  	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:120)
  	at java.lang.Thread.run(Thread.java:748)
  ]

    at __randomizedtesting.SeedInfo.seed([63AD7793D7B01433]:0)
    at org.junit.Assert.fail(Assert.java:88)
    at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount(ElasticsearchAssertions.java:261)
    at org.elasticsearch.search.basic.SearchWhileRelocatingIT$1.run(SearchWhileRelocatingIT.java:75)

```

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/basic/SearchWhileRelocatingIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/basic/SearchWhileRelocatingIT.java
@@ -8,6 +8,7 @@
 
 package org.elasticsearch.search.basic;
 
+import org.elasticsearch.action.NoShardAvailableActionException;
 import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
 import org.elasticsearch.action.index.IndexRequestBuilder;
 import org.elasticsearch.action.search.SearchPhaseExecutionException;
@@ -21,6 +22,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Stream;
 
 import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
@@ -65,10 +67,12 @@ public class SearchWhileRelocatingIT extends ESIntegTestCase {
                             while (stop.get() == false) {
                                 SearchResponse sr = client().prepareSearch().setSize(numDocs).get();
                                 if (sr.getHits().getTotalHits().value != numDocs) {
-                                    // if we did not search all shards but had no failures that is potentially fine
+                                    // if we did not search all shards but had no serious failures that is potentially fine
                                     // if only the hit-count is wrong. this can happen if the cluster-state is behind when the
                                     // request comes in. It's a small window but a known limitation.
-                                    if (sr.getTotalShards() != sr.getSuccessfulShards() && sr.getFailedShards() == 0) {
+                                    if (sr.getTotalShards() != sr.getSuccessfulShards() &&
+                                        Stream.of(sr.getShardFailures()).allMatch(
+                                            ssf -> ssf.getCause() instanceof NoShardAvailableActionException)) {
                                         nonCriticalExceptions.add("Count is " + sr.getHits().getTotalHits().value + " but " + numDocs +
                                             " was expected. " + formatShardStatus(sr));
                                     } else {
@@ -84,8 +88,7 @@ public class SearchWhileRelocatingIT extends ESIntegTestCase {
                             }
                         } catch (SearchPhaseExecutionException ex) {
                             // it's possible that all shards fail if we have a small number of shards.
-                            // with replicas this should not happen
-                            if (numberOfReplicas == 1 || ex.getMessage().contains("all shards failed") == false) {
+                            if (ex.getMessage().contains("all shards failed") == false) {
                                 throw ex;
                             }
                         }

